### PR TITLE
Feat: Add comprehensive JSDoc documentation for all functions and tests and README

### PR DESCRIPTION
**Pull request by juliopavila**
## Description:

This pull request adds comprehensive JSDoc documentation to various parts of the codebase, including internal functions, deployment scripts, and test cases. The documentation follows the JSDoc standard and provides detailed descriptions of parameters, return values, and exceptions for each function.

## Changes:
**•	Internal Functions:** Added detailed JSDoc comments to internal functions such as getBuildArtifact, resolvePaths, and searchFile, describing their purpose, parameters, and return values.

**•	Deployment Scripts:** Enhanced documentation for deployment-related functions, including deployKnownFactory, deployMastercopy, and deployProxy. Each function now includes comprehensive JSDoc comments to improve clarity and maintainability.

**•	Test Cases:** I updated the test files with JSDoc comments to describe the setup functions, individual tests, and expected outcomes. This includes documentation for tests in `SignatureChecker.spec.ts`, `IAvatar.spec.ts`, `Module.spec.ts`, and others.

## Context:

Adding detailed documentation helps improve code readability, makes it easier for new contributors to understand the codebase, and ensures that the purpose and usage of functions are clear. This aligns with best practices for maintaining a high-quality codebase.

## How Has This Been Tested?

The changes are purely documentation-based and do not alter any functionality. Existing tests have been run to ensure that the codebase remains stable.

### CODE REVIEW

File: src/encoding/encodeDeployProxy.ts [hunk 1]
```diff
@@ -1,18 +1,22 @@
...
+ *
+ * @param {Object} params - The function parameters.
+ * @param {string} [params.factory=factoryAddress] - The address of the factory contract.
+ * @param {string} params.mastercopy - The address of the mastercopy contract.
+ * @param {Object} params.setupArgs - The arguments for the setup function.
+ * @param {any[]} params.setupArgs.types - The types of the setup arguments.
+ * @param {any[]} params.setupArgs.values - The values of the setup arguments.
+ * @param {BigNumberish} params.saltNonce - The salt nonce used to predict the proxy address.
+ * @returns {TransactionRequest} The encoded transaction request.
+ */
```
Comment (cristovaoth) [inline src/encoding/encodeDeployProxy.ts:19]: Proposal:

/**
 * Encodes the transaction payload for deploying a Mod instance.
 *
 * The Mod instance is deployed as a minimal proxy through the `ZodiacModuleProxyFactory`.
 *
 * @param {Object} params - The function parameters.
 * @param {string} [params.factory=factoryAddress] - The address of the factory contract.
 * @param {string} params.mastercopy - The address of the mastercopy contract.
 * @param {Object} params.setupArgs - The mod setup args
 * @param {any[]} params.setupArgs.types - The types of the setup arguments.
 * @param {any[]} params.setupArgs.values - The values of the setup arguments.
 * @param {BigNumberish} params.saltNonce - The salt nonce used in the proxy deployment.
 * @returns {TransactionRequest} The encoded transaction request.
 */

File: src/encoding/encodeDeployVia2470Factory.ts [hunk 2]
```diff
@@ -5,6 +5,18 @@ import {
...
+ *
+ * @param {Object} params - The function parameters.
+ * @param {string} [params.factory=factoryAddress] - The address of the factory contract.
+ * @param {string} params.bytecode - The bytecode of the contract to deploy.
+ * @param {Object} params.constructorArgs - The constructor arguments for the contract.
+ * @param {any[]} params.constructorArgs.types - The types of the constructor arguments.
+ * @param {any[]} params.constructorArgs.values - The values of the constructor arguments.
+ * @param {BigNumberish} params.salt - The salt value used for CREATE2 deployment.
+ * @returns {TransactionRequest} The encoded transaction request.
+ */
```
Comment (cristovaoth) [inline src/encoding/encodeDeployVia2470Factory.ts:19]: Proposal:
/**
 * Encodes the transaction request for deploying a Singleton via ERC2470Factory.
 *
 * @param {Object} params - The function parameters.
 * @param {string} [params.factory=factoryAddress] - The address of the factory contract.
 * @param {string} params.bytecode - The bytecode of the contract to deploy.
 * @param {Object} params.constructorArgs - The constructor arguments for the Singleton that is about to be deployed.
 * @param {any[]} params.constructorArgs.types - The types of the constructor arguments.
 * @param {any[]} params.constructorArgs.values - Values corresponding to the constructor argument types.
 * @param {string} params.salt - A deployment salt that is internally passed to create2
 *
 * @returns {TransactionRequest} The encoded transaction request.
 */

File: src/encoding/encodeDeployViaNickFactory.ts [hunk 3]
```diff
@@ -2,6 +2,18 @@ import { AbiCoder, concat, TransactionRequest } from "ethers";
...
+ *
+ * @param {Object} params - The function parameters.
+ * @param {string} [params.factory=factoryAddress] - The address of the factory contract.
+ * @param {string} params.bytecode - The bytecode of the contract to deploy.
+ * @param {Object} params.constructorArgs - The constructor arguments for the contract.
+ * @param {any[]} params.constructorArgs.types - The types of the constructor arguments.
+ * @param {any[]} params.constructorArgs.values - The values of the constructor arguments.
+ * @param {BigNumberish} params.salt - The salt value used for CREATE2 deployment.
+ * @returns {TransactionRequest} The encoded transaction request.
+ */
```
Comment (cristovaoth) [inline src/encoding/encodeDeployViaNickFactory.ts:16]: Proposal:
/**
 * Encodes the transaction request for deploying a Singleton via Nick factory.
 *
 * @param {Object} params - The function parameters.
 * @param {string} [params.factory=factoryAddress] - The address of the factory contract.
 * @param {string} params.bytecode - The bytecode of the contract to deploy.
 * @param {Object} params.constructorArgs - The constructor arguments for the Singleton that is about to be deployed.
 * @param {any[]} params.constructorArgs.types - The types of the constructor arguments.
 * @param {any[]} params.constructorArgs.values - Values corresponding to the constructor argument types.
 * @param {string} params.salt - A deployment salt that is internally passed to create2
 *
 * @returns {TransactionRequest} The encoded transaction request.
 */

File: src/encoding/predictProxyAddress.ts [hunk 4]
```diff
@@ -9,6 +9,18 @@ import {
...
+ *
+ * @param {Object} params - The function parameters.
+ * @param {string} [params.factory=factoryAddress] - The address of the factory contract.
+ * @param {string} params.mastercopy - The address of the mastercopy contract.
+ * @param {Object} params.setupArgs - The arguments for the setup function.
+ * @param {any[]} params.setupArgs.types - The types of the setup arguments.
+ * @param {any[]} params.setupArgs.values - The values of the setup arguments.
+ * @param {BigNumberish} params.saltNonce - The salt nonce used for CREATE2 deployment.
+ * @returns {string} The predicted address of the proxy contract.
+ */
```
Comment (cristovaoth) [inline src/encoding/predictProxyAddress.ts:23]: Proposal:
/**
 * Predicts the address of a proxy contract deployed via ModuleProxyFactory.
 *
 * @param {Object} params - The function parameters.
 * @param {string} [params.factory=factoryAddress] - The address of the factory contract.
 * @param {string} params.mastercopy - The address of the mastercopy contract.
 * @param {Object} params.setupArgs - The arguments for the setup function.
 * @param {any[]} params.setupArgs.types - The types of the setup arguments.
 * @param {any[]} params.setupArgs.values - The values of the setup arguments.
 * @param {BigNumberish} params.saltNonce - The saltNonce used to internally derive the final create2 salt.
 * 
 * @returns {string} The predicted address of the proxy contract.
 */

File: src/encoding/predictSingletonAddress.ts [hunk 5]
```diff
@@ -2,25 +2,47 @@ import { AbiCoder, concat, getCreate2Address, keccak256 } from "ethers";
...
+ *
+ * @param {Object} params - The function parameters.
+ * @param {string} [params.factory=factoryAddress] - The address of the factory contract.
+ * @param {string} params.bytecode - The bytecode of the contract to deploy.
+ * @param {Object} params.constructorArgs - The constructor arguments for the contract.
+ * @param {any[]} params.constructorArgs.types - The types of the constructor arguments.
+ * @param {any[]} params.constructorArgs.values - The values of the constructor arguments.
+ * @param {BigNumberish} params.salt - The salt value used for CREATE2 deployment.
+ * @returns {string} The predicted address of the singleton contract.
+ */
```
Comment (cristovaoth) [inline src/encoding/predictSingletonAddress.ts:16]: Proposal:
/**
 * Predicts the address of a proxy contract deployed via SingletonFactory.
 *
 * Note: The calculation method is the same regardless of whether the factory is Nick or EIP2470.
 *
 * @param {Object} params - The parameters for predicting the address.
 * @param {string} [params.factory=erc3460FactoryAddress] - The address of the factory contract.
 * @param {string} params.bytecode - The bytecode of the contract to deploy.
 * @param {Object} params.constructorArgs - The constructor arguments for the contract.
 * @param {any[]} params.constructorArgs.types - The types of the constructor arguments.
 * @param {any[]} params.constructorArgs.values - Values corresponding to the constructor argument types.
 * @param {string} params.salt - A deployment salt that is internally passed to create2
 *
 * @returns {string} The predicted address of the singleton contract.
 */

File: src/deploy/deployFactories.ts [hunk 6]
```diff
@@ -22,7 +22,18 @@ import waitForTransaction from "./waitForTransaction";
...
 import { EIP1193Provider } from "../types";
 
-export default async function ({ provider }: { provider: EIP1193Provider }) {
+/**
+ * Deploys known factories and the mastercopy contract.
+ *
+ * @param {Object} params - The function parameters.
+ * @param {EIP1193Provider} params.provider - The EIP1193 compliant provider to interact with the blockchain.
+ * @returns {Promise<void>} A promise that resolves once the factories and mastercopy are deployed.
+ */
```
Comment (cristovaoth) [inline src/deploy/deployFactories.ts:31]: Proposal:

/**
 * Deploys all factories within a specified network.
 *
 * This function deploys the following factories:
 * - `NickSingletonFactory`
 * - `ERC2470SingletonFactory`
 * - `ZodiacModuleProxyFactory`
 *
 * The first two factories, `NickSingletonFactory` and `ERC2470SingletonFactory`, use known presigned transactions for deployment.
 * The `ZodiacModuleProxyFactory` is deployed as a singleton via the `ERC2470SingletonFactory`.
 *
 * If all factories are already deployed, this function performs no operation.
 *
 * Note: Typically, these factories are already deployed across networks. This function is mostly useful for test setups.
 *
 * @param {Object} params - The function parameters.
 * @param {EIP1193Provider} params.provider - The EIP1193 compliant provider to interact with the blockchain.
 *
 * @returns {Promise<void>} A promise that resolves once the factories are deployed.
 */

File: src/deploy/deployMastercopy.ts [hunk 7]
```diff
@@ -13,6 +9,20 @@ import waitForTransaction from "./waitForTransaction";
...
+ * @param {string} [params.factory=erc2470FactoryAddress] - The address of the factory contract to use.
+ * @param {string} params.bytecode - The bytecode of the contract to deploy.
+ * @param {Object} params.constructorArgs - The constructor arguments for the contract.
+ * @param {any[]} params.constructorArgs.types - The types of the constructor arguments.
+ * @param {any[]} params.constructorArgs.values - The values of the constructor arguments.
+ * @param {string | number | bigint} params.salt - The salt value used for CREATE2 deployment.
+ * @param {EIP1193Provider} params.provider - The EIP1193 compliant provider to interact with the blockchain.
+ * @param {function} [params.onStart] - Optional callback function to call when the deployment starts.
+ * @returns {Promise<{ address: string; noop: boolean }>} The address of the deployed mastercopy contract and a noop flag indicating if the deployment was a no-operation.
+ */
```
Comment (cristovaoth) [inline src/deploy/deployMastercopy.ts:25]: Proposal:

/**
 * Deploys a Mastercopy via a SingletonFactory
 *
 * @param {Object} params - The function parameters.
 * @param {string} [params.factory=erc2470FactoryAddress] - The address of the factory to use.
 * @param {string} params.bytecode - The bytecode of the contract to deploy.
 * @param {Object} params.constructorArgs - The constructor arguments for the contract.
 * @param {any[]} params.constructorArgs.types - The types of the constructor arguments.
 * @param {any[]} params.constructorArgs.values - The values of the constructor arguments.
 * @param {string | number | bigint} params.salt - The salt value used for CREATE2 deployment.
 * @param {EIP1193Provider} params.provider - The EIP1193 compliant provider to interact with the blockchain.
 * @param {function} [params.onStart] - Optional callback function to call when the deployment starts.
 *
 * @returns {Promise<{ address: string; noop: boolean }>} The address of the deployed mastercopy contract and a noop flag indicating if the deployment was a no-operation.
 */

File: src/deploy/deployProxy.ts [hunk 8]
```diff
@@ -1,10 +1,21 @@
...
+ *
+ * @param {Object} params - The function parameters.
+ * @param {string} params.mastercopy - The address of the mastercopy contract.
+ * @param {Object} params.setupArgs - The arguments for the setup function.
+ * @param {any[]} params.setupArgs.types - The types of the setup arguments.
+ * @param {any[]} params.setupArgs.values - The values of the setup arguments.
+ * @param {string | number | bigint} params.saltNonce - The salt nonce used to predict the proxy address.
+ * @param {EIP1193Provider} params.provider - The EIP1193 compliant provider to interact with the blockchain.
+ * @returns {Promise<{ address: string; noop: boolean }>} The address of the deployed proxy contract and a noop flag indicating if the deployment was a no-operation.
+ */
```
Comment (cristovaoth) [inline src/deploy/deployProxy.ts:18]: Proposal:

/**
 * Deploys a Mod instance as a MinimalProxy via ZodiacModProxyFactory
 *
 * @param {Object} params - The function parameters.
 * @param {string} params.mastercopy - The address of the mastercopy contract.
 * @param {Object} params.setupArgs - The arguments for the setup function.
 * @param {any[]} params.setupArgs.types - The types of the setup arguments.
 * @param {any[]} params.setupArgs.values - The values of the setup arguments.
 * @param {string | number | bigint} params.saltNonce - The salt nonce used to predict the proxy address.
 * @param {EIP1193Provider} params.provider - The EIP1193 compliant provider to interact with the blockchain.
 *
 * @returns {Promise<{ address: string; noop: boolean }>} The address of the deployed proxy contract and a noop flag indicating if the deployment was a no-operation.
 */

**Comment by cristovaoth** [comment]
The README still lacks some detailed information about the exported Solidity contracts, but the SDK functions now have more detailed descriptions.

**Review by samepant — APPROVED** [review]

**Pull request merged**